### PR TITLE
Introducing Structured Data Types

### PR DESCRIPTION
**Pull request by ftomassetti**
Ths PR introduces Structured Data Types in the M3, M2, and M1.

Todo:
- [x] Write tests showing serialization of SDTs (it should work out of the box)
- [x] Support serialization of Structured Data Type Instances

Note that this should wait for #167 or #170 to be merged, as they move the LW version to 2024.1, which is the version where Structured Data Types are introduced. Until then, some tests will fail because LionCore does not match the expectation (as we kept version 2023.1 and introduced stuff not belonging to that version). **In other words, while this can be reviewed it cannot yet be merged.**

### CODE REVIEW

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java [hunk 1]
```diff
@@ -72,7 +73,8 @@ public Object getPropertyValue(Property property) {
    * This internal method uses a property name and not a property or the property id because of a
    * circular dependency problem present for nodes representing M3 elements.
    */
-  protected <V> V getPropertyValue(String propertyName, Class<V> clazz, V defaultValue) {
+  protected <V> V getPropertyValue(
```
Comment (enikao) [inline core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java:76]: mark return and `defaultValue` `@Nullable`?

File: core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java [hunk 2]
```diff
@@ -176,6 +187,14 @@ public static Language getInstance() {
       annotation.addFeature(
           Reference.createMultiple("implements", iface, "-id-Annotation-implements"));
 
+      structuredDataType.setExtendedConcept(dataType);
+      structuredDataType.addFeature(
```
Comment (enikao) [inline core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java:191]: Is this _required_ by default?

File: core/src/test/java/io/lionweb/lioncore/java/serialization/MyNodeWithStructuredDataType.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+package io.lionweb.lioncore.java.serialization;
+
+import io.lionweb.lioncore.java.language.*;
+import io.lionweb.lioncore.java.model.ClassifierInstanceUtils;
+import io.lionweb.lioncore.java.model.StructuredDataTypeInstance;
+import io.lionweb.lioncore.java.model.impl.DynamicNode;
+
+public class MyNodeWithStructuredDataType extends DynamicNode {
```
Comment (enikao) [inline core/src/test/java/io/lionweb/lioncore/java/serialization/MyNodeWithStructuredDataType.java:8]: Maybe add an example (and tests) for nested SDTs, and SDTs containing enums.

In C#, I use the following example:

```
enum Currency
  EUR
  GBP

sdt Decimal
  int: Integer
  frac: Integer

sdt Amount
  value: Decimal
  currency: Currency
  digital: Boolean
```

File: core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java [hunk 4]
```diff
@@ -13,16 +13,21 @@ public interface HasFeatureValues {
    *
    * <p>Should this return a String?
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java:14]: Right, fixing it

File: core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java [hunk 5]
```diff
@@ -13,16 +13,21 @@ public interface HasFeatureValues {
...
    */
-  Object getPropertyValue(Property property);
+  @Nullable
+  Object getPropertyValue(@Nonnull Property property);
 
-  /** If the value is not compatible with the type of the property, the exception */
-  void setPropertyValue(Property property, Object value);
+  /**
+   * If the value is not compatible with the type of the property, the exception
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java:20]: Right, doing as suggested

File: core/src/main/java/io/lionweb/lioncore/java/model/StructuredDataTypeInstance.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package io.lionweb.lioncore.java.model;
+
+import io.lionweb.lioncore.java.language.Field;
+import io.lionweb.lioncore.java.language.StructuredDataType;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/** This a represent in instance of Structured Data Type. */
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/StructuredDataTypeInstance.java:8]: Fixing it, thank you

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      throw new IllegalArgumentException(
+          "Invalid field for StructuredDataType " + getStructuredDataType());
+    }
+    return fieldValues.get(field);
+  }
+
+  @Override
+  public void setFieldValue(@Nonnull Field field, @Nullable Object value) {
+    Objects.requireNonNull(field, "field should not be null");
+    if (field.getID() == null) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java:75]: ID, as I had a situation when the ID was null and I got an NPE down the line

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java [hunk 8]
```diff
@@ -61,7 +61,8 @@ public Containment getContainmentFeature() {
   }
 
   @Override
-  public Object getPropertyValue(Property property) {
+  public Object getPropertyValue(@Nonnull Property property) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java:64]: Right, done

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java (hunk 1 above)
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/M3Node.java:76]: Done, thank you

File: core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java (hunk 2 above)
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java:191]: Do you mean `fields`? No it is multiple and multiple is not required by default, it is optional. So the multiplicity is 0..*

File: core/src/test/java/io/lionweb/lioncore/java/serialization/MyNodeWithStructuredDataType.java (hunk 3 above)
Comment (ftomassetti) [reply core/src/test/java/io/lionweb/lioncore/java/serialization/MyNodeWithStructuredDataType.java:8]: I added such example and two tests covering it

File: core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java [hunk 9]
```diff
@@ -32,7 +37,7 @@ public interface HasFeatureValues {
    * <p>If the child has not a Concept compatible with the target of the Containement, the exception
    * IllegalArgumentException will be thrown.
    */
-  void addChild(Containment containment, Node child);
+  void addChild(@Nonnull Containment containment, @Nullable Node child);
```
[the remark this answers is not in the document]
Comment (enikao) [reply core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java:36]: Broken models are ok, but I can't see a case where adding a `null` child makes sense.

File: core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java (hunk 2 above)
Comment (enikao) [reply core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java:191]: in the spec M3 it's required

File: core/src/test/java/io/lionweb/lioncore/java/serialization/SerializationOfStructuredDataTypeInstancesTest.java [hunk 10]
```diff
@@ -158,4 +161,54 @@ public void deserializeStructuredDataTypeInstance() {
...
+            .serialize(
+                MyNodeWithAmount.CURRENCY.getID(),
+                new EnumerationValueImpl(MyNodeWithAmount.CURRENCY.getLiterals().get(0)));
+    assertEquals("euro", currencySerialized);
+    String amountSerialized =
+        jsonSerialization
+            .getPrimitiveValuesSerialization()
+            .serialize(MyNodeWithAmount.AMOUNT.getID(), amount);
+    assertEquals(
+        "{\"value\":\"{\\\"int\\\":\\\"2\\\",\\\"frac\\\":\\\"3\\\"}\",\"currency\":\"euro\",\"digital\":\"true\"}",
```
Comment (enikao) [inline core/src/test/java/io/lionweb/lioncore/java/serialization/SerializationOfStructuredDataTypeInstancesTest.java:200]: Is this according to spec? Decoding the escapes is hard without tooling.
Spec says to encode the whole json tree once, not each sub-tree: https://github.com/LionWeb-io/specification/blob/main/serialization/serialization.adoc#structured-datatype

File: core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/HasFeatureValues.java]: Ok, I have changed it to Nonnull here and in all the classes implementing it

File: core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java (hunk 2 above)
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/self/LionCore.java:191]: Oh, I see. I have changed it accordingly

File: core/src/main/java/io/lionweb/lioncore/java/serialization/PrimitiveValuesSerialization.java [hunk 11]
```diff
@@ -147,28 +170,28 @@ public Object deserialize(DataType dataType, String serializedValue, boolean isR
...
-      }
-      return sdtInstance;
+      return deserializeSDT(dataTypeID, jo);
     } else {
       throw new IllegalArgumentException(
           "Unable to deserialize primitive values of type " + dataType);
     }
   }
 
+  private JsonObject serializeSDT(@Nullable StructuredDataTypeInstance structuredDataTypeInstance) {
```
Comment (enikao) [inline core/src/main/java/io/lionweb/lioncore/java/serialization/PrimitiveValuesSerialization.java:181]: Why is the parameter `@Nullable`? Would throw an NPE in line 183

**Comment by enikao** [comment]
Did you look at these comments?

image

**Comment by ftomassetti** [comment]
> Did you look at these comments?
> 
> image

I did not, sorry, I missed them. I will look into them right now

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java [hunk 12]
```diff
@@ -47,12 +47,15 @@ public Object getPropertyValue(@Nonnull Property property) {
...
 
   @Override
-  public void setPropertyValue(@Nonnull Property property, Object value) {
+  public void setPropertyValue(@Nonnull Property property, @Nullable Object value) {
     Objects.requireNonNull(property, "Property should not be null");
     if (!getClassifier().allProperties().contains(property)) {
       throw new IllegalArgumentException(
           "Property " + property + " is not belonging to classifier " + getClassifier());
     }
+    if (property.getID() == null) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java:56]: Keys have the particularity of being stable across versions, but besides that my understanding is that they have no other special role. In general, both the key and the id should be defined and be unique within the features of a node, so both could be used as keys in a map of features. I tend to use IDs and not keys precisely because IDs are unique _also across versions_, so they permit the identification of features more uniquely. 

In this particular case, either the id or the key would work. I think there is a slight advantage in using the id because I could represent a state in which the node is being migrated, and for an instant, it has values for the same property in two different versions of the language. In a similar situation I could store both the value for the "old" property and the "new" property, while if I were using the key I would not be able to recognize the case. This seems admittedly a corner case.

In other words, I do not think the specs indicate that internally we should use organize features by ids or keys, as both should be unique. I would tend to use ids and not keys because I think they can be (very marginally) better because they are more unique, and because when in doubt I would let the status quo prevail. Does it make sense?

I can definitely move the check up.

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  private Map<Field, Object> fieldValues = new HashMap<>();
+
+  public DynamicStructuredDataTypeInstance(@Nonnull StructuredDataType structuredDataType) {
+    Objects.requireNonNull(structuredDataType, "structuredDataType should not be null");
+    this.structuredDataType = structuredDataType;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof StructuredDataTypeInstance)) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java:27]: Good idea, I will do so

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+  @Override
+  public @Nonnull StructuredDataType getStructuredDataType() {
+    return structuredDataType;
+  }
+
+  @Override
+  public @Nullable Object getFieldValue(@Nonnull Field field) {
+    Objects.requireNonNull(field, "field should not be null");
+    if (field.getID() == null) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java:62]: See comment above: I think this is an implementation choice, and I have a (slight) preference for ID, but I think that both would work here as both should be unique

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicStructuredDataTypeInstance.java:75]: See also the comment above regarding key vs ID for differentiation features

File: core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (enikao) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java]: Disclaimer: I think the key vs. id debate has a large "philosophical" part, so take my arguments in that view.

In another (very corner) case one could imagine a hard-coded language that does not have ids for its elements -- we don't require language elements to have ids (kind-of, I'm pretty sure all implementations still require it to keep the "everything is a node" mantra).

I think it gets a lot more interesting when we start actually using the _version_ part of a meta-pointer: `language` (key) + `version` (of the language) uniquely identify the language; within that, the `key` uniquely identifies the element within the language. 
Element key and language key have to match perfectly, otherwise they don't make sense.

Version is a different matter. We explicitly separated it from language key to allow some "softness" in matching. Example: In my system I only know language `A` in version `x` (written as `A@x`), but try to deserialize an instance of `A` in version `y`. Then I could still use the language and elements of `A@x` and either be lucky that they are compatible, or at least deserialize as much as possible.

I imagine then it might get important to use keys instead of ids, because they might be applicable to the same language in a different version -- which is the desired effect. It's kind of the same argument as you have, but with a very different interpretation (-:

[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicClassifierInstance.java]: On the key vs. id debate I agree it can get "philosophical" and I think it has impact in at least two points:
- In this class
- In PrimitiveValuesSerialization

In this class, I do not feel strongly, so if you prefer we use keys instead of IDs, that works for me. I have implemented this change in the last commit.

In PrimitiveValuesSerialization instead I feel strongly, as I think that we may want to specify different ways to serialize different versions of a certain primitive type. So there I use ids and not keys.

File: core/src/main/java/io/lionweb/lioncore/java/serialization/PrimitiveValuesSerialization.java [hunk 15]
```diff
@@ -72,6 +72,29 @@ public PrimitiveValuesSerialization registerSerializer(
     return this;
   }
 
+  private StructuredDataTypeInstance deserializeSDT(String dataTypeID, JsonObject jo) {
+    StructuredDataType sdt = strucuturesDataTypesByID.get(dataTypeID);
+    DynamicStructuredDataTypeInstance sdtInstance = new DynamicStructuredDataTypeInstance(sdt);
+    for (Field field : sdt.getFields()) {
+      if (jo.has(field.getName())) {
```
Comment (enikao) [inline core/src/main/java/io/lionweb/lioncore/java/serialization/PrimitiveValuesSerialization.java:79]: According to spec, we use the field's key in the JSON structure, not it's name.

Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/serialization/PrimitiveValuesSerialization.java]: Right, it should be now fixed

**Comment by ftomassetti** [comment]
Rebased after merging the PR to support multiple LW Versions

**Comment by ftomassetti** [comment]
At this stage a single test should be failing because of https://github.com/LionWeb-io/specification/issues/324. In practice here we have Field implementing INamed, which is not the case in the specs

**Comment by ftomassetti** [comment]
I was wrong about https://github.com/LionWeb-io/specification/issues/324, so I changed Field not to implement INamed directly and now tests should all pass

**Comment by ftomassetti** [comment]
I think we covered all the open comments, and the PR has been open for two weeks, so I will go ahead and merge it. In this way we can hit the deadline of completing this release by the end of the year. Of course we can cover any additional issues in further PRs. Thank you!
